### ISSUE TITLE
Auto set header from cookie

### ISSUE THREAD

**Issue by MJegorovas** [ktorio/ktor#1972]
**Subsystem**
Client-Cookies(?)

**Is your feature request related to a problem? Please describe.**
A server expects a specific header which I can get from cookies. Now, because the header could change anytime I need some logic that would set the newest header value for every request. I could add 'headers' parameter to every 'HttpRequestBuilder', but that would introduce a lot of boilerplate.

**Describe the solution you'd like**
Best solution would be for ktor to have some logic to automatically add header from cookies like some [JavaScript frameworks](https://laravel.com/docs/master/csrf#csrf-x-xsrf-token) do, but I don't know how it should be done. Second best would be an example how to make interceptor to add header to every 'HttpStatement' (couldn't find any examples for client that would also be great for getting cookie value from reponse).

**Comment by wyaeld** [comment]
There already exists

What you want to do is configure your HttpClient connection with an Authentication feature, using a very simple AuthProvider that you extend/implement yourself.

That interface has a function on it for `addRequestHeaders`.  That allow you complete control to add custom logic

**Comment by MJegorovas** [comment]
Thanks for answer wyaeld

I don't think your solution is any different than implementing custom interceptor to add header to request (I did it this way). Also, I still did not find a way how to get cookie from cookie storage (by installing 'AcceptAllCookies') by myself, so had to do a workaround for this too. Feel free to comment if u know a way to do that.

From my research I found that cookies are written to headers on request 'Send' phase (not sure where exactly), but custom interceptor on this phase wouldn't work because request first would be executed and only then my interceptor would be applied.
